EmbeddedScripts.CSharp/CodeCompiler.cs

```diff
@@ -0,0 +1,58 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using EmbeddedScripts.Shared;
+
+namespace EmbeddedScripts.CSharp
+{
+    internal class CodeCompiler
+    {
+        public CodeCompiler(string code, Container container)
+        {
+            Container = container;
+            Compilation = Compile(code);
+        }
+
+        public CSharpCompilation Compilation { get; set; }
+
+        private CSharpCompilation Compile(string code)
+        {
+            var syntaxTree = CSharpSyntaxTree.ParseText(code);
+
+            var assemblyName = Path.GetRandomFileName();
+
+            var references = GenerateMetadataReferences(Container);
+
+            var opts = new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary);
+
+            return CSharpCompilation.Create(
+                assemblyName,
+                new[] { syntaxTree },
+                references,
+                opts
+            );
+        }
+
+        private List<MetadataReference> GenerateMetadataReferences(Container container)
```

Comment:
Sometimes List<T>, sometimes IEnemurable<T>. I don't use List<T> features, is it a good idea to switch to IEnumerable<T>?

Reply:
In general you should return List when you are planning to modify the collection, return IEnumerable if you are planning to enumerate it in foreach loops only, return Array if you need the data as array.